keps/0005-cluster-resources-for-crds.md

```diff
@@ -85,25 +88,49 @@ should probably be created
 
 OperatorVersions require the existance of CRDs that are not controlled by Kudo (e.g. ETCD Operator) and require those to be installed when FV is enabled.
 
-### Implementation Details/Notes/Constraints [optional]
+### Implementation Details/Notes/Constraints
 
+The installation of an Operator Version can run a plan specified by the Operator Version spec, when the `installPlan` value is specified in `operator.yaml`.  
+The value shall match a defined `plan` in the `plans` spec in the OperatorVersion. As with the Instance Status properties, the OperatorVersion will contain
 
+```golang
+	PlanStatus       map[string]PlanStatus `json:"planStatus,omitempty"`
+	AggregatedStatus AggregatedStatus      `json:"aggregatedStatus,omitempty"`
+```
 
-### Risks and Mitigations
+to contain the status information on the execution of the `installPlan`. The runtime logic for executing the plan on OperatorVersion install will be the same core logic that is used in the Instance plan execution.
+
+#### Instance Status update
+
+If Instances of an OperatorVersion are created before the Install plan of the OperatorVersion is successful, the Instance's will remian in a `PENDING` status until the OperatorVersion is successful.
+
+#### Object Owners
 
+The objects created by the `installPlan` may exist on the cluster before the OperatorVersion is installed. This may be a result of another OperatorVersion requiring the same objects, or the objects being managed externally to KUDO. If so, the OperatorVersion will be appended to the `ownerReferences`. If the object does not exist, then KUDO will create the object with the OperatorVersion as the `ownerReference`, and additionally add an annotation to the object that it was created by KUDO: `kudo.dev/managed: true`. This will be used during cleanup to identify whether the object was created by KUDO or not.
```

Comment:
On the topic of owning object and `ownerReferences`. To quote the [GC docs](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/):
```
Cross-namespace owner references are disallowed by design. This means: 
1) Namespace-scoped dependents can only specify owners in the same namespace, and owners that are cluster-scoped. 
2) Cluster-scoped dependents can only specify cluster-scoped owners, but not namespace-scoped owners.
```
So e.g. a **CRD** (not a CR) is a **cluster-scoped** resource and can't/shall not be owned by a **namespaced OV or O**. The fact that we do it right now and the fact that GC-ing them works is non-deterministic GC behavior and is discouraged.  

So an _Instance_ can install a CRD but it won't/can't be GC-ed when the instance is removed. I guess we'll have to make better use of the `cleanup` plans 🤷‍♂

Reply:
Good find on this.  It sounds like there's nothing stopping us form using the `ownerReference` field to capture the concept that the CRD (cluster scoped) was installed to support an OV.  The caveat is that we can't expect the Garbage collector to do the cleanup for us.

Since we'll be doing this ourselves, does it make more sense to have an annotation for each OV that's using it?
```yaml
metadata:
  labels:
    kudo.dev/managed: true
    kudo.dev/owned: default/kafka-0.1.3
    kudo.dev/owned: otherns/redis-1.5
...
```

I always find labels easier to read and search with.

Reply:
Multiple Labels with the same key are not allowed. You can send yaml like this, but only one of them get persisted.

Apart from that I would agree that we should use labels. I'm not sure if we can rely on ownerReferences be settable across namespaces. Even if it works now, it might get denied in a future K8s version when validation is tighter.

Maybe finalizers would work?
```yaml
finalizer:
  - kudo_owner: default/kafka-0.1.3
  - kudo_owner: otherns/redis-1.5
```

As we have to rely on the Controller anyway for cleanup, that might be a usable way?

Reply:
Duh.  Absolutely correct.  
One tag setup might better correspond to the Tags used by EKS for cluster autoscaling:

https://docs.aws.amazon.com/eks/latest/userguide/cluster-autoscaler.html

Which would have our tags look like:
```yaml
kudo.dev/operator-version/default/kafka-0.1.3/owned: true
kudo.dev/operator-version/default/redis-1.5.0/owned: true
```

I lean towards Labels over annotations/finalizers to allow for answering questions like "Which OVs use this CRD?" and "Which CRDs are owned by this OV?"

Reply:
Yeah, that would work.
We should add finalizers for each OV as well, though, to prevent someone from deleting the installed CRDs